File: backend/app/schemas/library.py
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime, date
from typing import Optional

class BookBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    author: str = Field(..., min_length=1, max_length=255)
    isbn: str = Field(..., min_length=5, max_length=20)
    category: str = Field(..., min_length=2, max_length=50)
    copies_total: int = Field(1, ge=0)

class BookCreate(BookBase):
    pass

class BookResponse(BookBase):
    id: str
    copies_available: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class BorrowBase(BaseModel):
    book_id: str
    student_id: str
    issue_date: date
    due_date: date

class BorrowCreate(BorrowBase):
    pass

class BorrowResponse(BorrowBase):
    id: str
    return_date: Optional[date] = None
    fine_amount: float
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class ReturnRequest(BaseModel):
    return_date: date

File: apps/web/src/api/library.ts
import { apiFetch } from "../lib/apiClient";

export interface Book {
  id: string;
  title: string;
  author: string;
  isbn: string;
  category: string;
  copies_total: number;
  copies_available: number;
  created_at: string;
}

export interface BookBorrow {
  id: string;
  book_id: string;
  student_id: string;
  issue_date: string;
  due_date: string;
  return_date?: string;
  fine_amount: number;
  created_at: string;
}

export async function apiGetBooks(skip = 0, limit = 50): Promise<Book[]> {
  return apiFetch<Book[]>(`/api/v1/books?skip=${skip}&limit=${limit}`);
}

export async function apiGetBook(id: string): Promise<Book> {
  return apiFetch<Book>(`/api/v1/books/${id}`);
}

export async function apiCreateBook(data: Partial<Book>): Promise<Book> {
  return apiFetch<Book>("/api/v1/books", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
}

export async function apiBorrowBook(bookId: string, studentId: string, days = 14): Promise<BookBorrow> {
  const issue = new Date().toISOString().split("T")[0];
  const due = new Date(Date.now() + days * 24 * 60 * 60 * 1000).toISOString().split("T")[0];
  
  return apiFetch<BookBorrow>(`/api/v1/books/${bookId}/borrow`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      book_id: bookId,
      student_id: studentId,
      issue_date: issue,
      due_date: due,
    }),
  });
}

export async function apiReturnBook(borrowId: string): Promise<BookBorrow> {
  const today = new Date().toISOString().split("T")[0];
  return apiFetch<BookBorrow>(`/api/v1/borrows/${borrowId}/return`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ return_date: today }),
  });
}

export async function apiGetStudentBorrows(studentId: string): Promise<BookBorrow[]> {
  return apiFetch<BookBorrow[]>(`/api/v1/borrows/student/${studentId}`);
}
